src/ripple/unl/tests/BasicNetwork.h

```diff
@@ -57,132 +60,176 @@ class BasicNetwork
+                    boost::intrusive::constant_time_size<false>>::type;
+        
+        using by_when_set = typename
+            boost::intrusive::make_multiset<msg,
+                boost::intrusive::constant_time_size<false>>::type;
 
-    using links_type =
-        boost::container::flat_map<Peer*, link_type>;
+        std::unordered_map<Peer*, by_to_list> by_to_;
+        std::unordered_map<Peer*, by_from_list> by_from_;
+        by_when_set by_when_;
+        qalloc alloc_;
 
-    class link_transform
-    {
     public:
-        using argument_type =
-            typename links_type::value_type;
+        using iterator =
+            typename by_when_set::iterator;
 
-        class result_type
-        {
-        public:
-            Peer& to;
-            bool inbound;
-
-            result_type (result_type const&) = default;
-
-            result_type (BasicNetwork& net, Peer& from,
-                    Peer& to_, bool inbound_)
-                : to(to_)
-                , inbound(inbound_)
-                , net_(net)
-                , from_(from)
-            {
-            }
+        explicit
+        queue_type (qalloc const& alloc);
 
-            /** Disconnect this link.
+        ~queue_type();
 
-                Effects:
+        bool
+        empty() const;
 
-                    The connection is removed at both ends.
-            */
-            void
-            disconnect() const
-            {
-                net_.disconnect(from_, to);
-            }
+        iterator
+        begin();
 
-        private:
-            BasicNetwork& net_;
-            Peer& from_;
-        };
+        iterator
+        end();
 
-        link_transform (BasicNetwork& net, Peer& from)
-            : net_(net)
-            , from_(from)
-        {
-        }
+        template <class Handler>
+        typename by_when_set::iterator
+        emplace (Peer* from, Peer* to,
+            time_point when, Handler&& h);
 
-        result_type const
-        operator()(argument_type const& v) const
-        {
-            return result_type(net_, from_,
-                *v.first, v.second.inbound);
-        }
+        void
+        erase (iterator iter);
 
-    private:
-        BasicNetwork& net_;
-        Peer& from_;
+        void
+        remove (Peer* from, Peer* to);
     };
 
-    struct peer_transform
+    struct link_type
     {
-        using argument_type =
-            typename links_type::value_type;
-
-        using result_type = Peer&;
+        bool inbound;
+        duration delay;
 
-        result_type
-        operator()(argument_type const& v) const
+        link_type (bool inbound_,
+                duration delay_)
+            : inbound (inbound_)
+            , delay (delay_)
         {
-            return *v.first;
         }
     };
 
-    msg_list msgs_;
+    using links_type =
+        boost::container::flat_map<Peer*, link_type>;
+
+    class link_transform;
+
+    qalloc alloc_;
```

Comment:
Yes, we want the allocator to be shared with as many class instances for which the behavior of `qalloc` is appropriate (objects are destroyed in approximately the same order which they are created). The member is public so that test code which uses `BasicNetwork` can use the allocator for containers or objects for which it is appropriate.